onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/embed_layer_norm_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "float.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+// EmbedLayerNorm supports limited data types.
+static std::vector<std::string> supported_data_types{"tensor(float16)", "tensor(float)", "tensor(int32)"};
+
+static bool IsSupportedDataType(const Node& node) {
+  for (const auto& input_arg : node.InputDefs()) {
+    if (std::find(supported_data_types.begin(), supported_data_types.end(),
+                  *(input_arg->Type())) == supported_data_types.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+Embed Layer Normalization will fuse embeddings and mask processing into one node : 
+The embeddings before conversion:
+  (input_ids) -------->  Gather ----------+       (segment_ids)
+    |                                    |            |
+    |                                    v            v
+    +--> Shape --> Expand -> Gather---->Add         Gather
+    |                ^                   |            |
+    |                |                   v            v
+    +---(optional graph)               SkipLayerNormalization
+
+*/
+Status EmbedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto node_index : node_topology_list) {
+    auto* p_skip_ln = graph.GetNode(node_index);
+    if (p_skip_ln == nullptr)
+      continue;  // we removed the node as part of an earlier fusion
+
+    Node& skip_ln_node = *p_skip_ln;
+    ORT_RETURN_IF_ERROR(Recurse(skip_ln_node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(skip_ln_node, "SkipLayerNormalization", {1}, kMSDomain) ||
+        !graph_utils::IsSupportedProvider(skip_ln_node, GetCompatibleExecutionProviders()) ||
+        !IsSupportedDataType(skip_ln_node)) {
```

Comment:
SkipLayerNormalization op only supports float and float16. IsSupportedDataType allows float/float16/double. So, there is no need to call IsSupportedDataType(). It will always be True.

Reply:
When Yufeng update SkipLayerNormalization to add an optional input of bias, we shall also check that there is no bias input here.

Reply:
When SkipLayerNormalization has an optional input (see yufeng's pull request https://github.com/microsoft/onnxruntime/pull/2453), there is also need to check that bias input does not exist. It depends on whether that request is checked in first.

Reply:
Got it. Will add the bias input check. So if the SkipLayerNormalization doesn't have the bias input, it should skip?

Reply:
It is opposite: when there is bias input, it should skip.